src/sources/open_coil.cpp

```diff
@@ -0,0 +1,333 @@
+      double face_weight = FTr->Face->Weight();
+      double val = 0.0;
+      FTr->Elem1->SetIntPoint(&eip);
+      elem.CalcVShape(*FTr->Elem1, dshape);
+      mfem::CalcOrtho(FTr->Face->Jacobian(), normal_vec);
+      val += dshape.InnerProduct(normal_vec, local_dofs) / face_weight;
+
+      // Measure the area of the boundary
+      area += ip.weight * face_weight;
+
+      // Integrate alpha * n.Grad(x) + beta * x
+      flux += val * ip.weight * face_weight;
+    }
+  }
+
+  double total_flux;
+  MPI_Allreduce(&flux, &total_flux, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+
+  return total_flux;
+}
+
+void SubdomainToArray(const std::vector<hephaestus::Subdomain> &sd,
+                      mfem::Array<int> &arr) {
+  arr.DeleteAll();
+  for (auto s : sd)
+    arr.Append(s.id);
+}
+
+void SubdomainToArray(const hephaestus::Subdomain &sd, mfem::Array<int> &arr) {
+  arr.DeleteAll();
+  arr.Append(sd.id);
+}
+
+template <typename T> void ifDelete(T *ptr) {
+  if (ptr != nullptr)
+    delete ptr;
+}
+
+void inheritBdrAttributes(const mfem::ParMesh *parent_mesh,
+                          mfem::ParSubMesh *child_mesh) {
+
+  int face, ori, att;
+  auto map = child_mesh->GetParentToSubMeshFaceIDMap();
+
+  for (int bdr = 0; bdr < parent_mesh->GetNBE(); ++bdr) {
+
+    parent_mesh->GetBdrElementFace(bdr, &face, &ori);
+    if (map[face] != -1) {
+      att = parent_mesh->GetBdrAttribute(bdr);
+      auto *new_elem = child_mesh->GetFace(map[face])->Duplicate(child_mesh);
+      new_elem->SetAttribute(att);
+      child_mesh->AddBdrElement(new_elem);
+    }
+  }
+
+  child_mesh->FinalizeTopology();
+  child_mesh->Finalize();
+  child_mesh->SetAttributes();
+}
+
+void cleanDivergence(hephaestus::GridFunctions *gridfunctions,
+                                     std::string J_name, std::string V_name,
+                                     hephaestus::BCMap *bc_map) {
+
+  hephaestus::InputParameters pars;
+  hephaestus::FESpaces fes;
+
+  pars.SetParam("VectorGridFunctionName", J_name);
+  pars.SetParam("ScalarGridFunctionName", V_name);
+  hephaestus::HelmholtzProjector projector(pars);
+  projector.Project(*gridfunctions, fes, *bc_map);
+}
+
+/////////////////////////////////////////////////////////////////////
+
+OpenCoilSolver::OpenCoilSolver(const hephaestus::InputParameters &params,
+                               const mfem::Array<int> &coil_dom,
+                               const std::pair<int, int> electrodes)
+    : J_gf_name_(params.GetParam<std::string>("SourceName")),
+      V_gf_name_(params.GetParam<std::string>("PotentialName")),
+      I_coef_name_(params.GetParam<std::string>("IFuncCoefName")),
+      coil_domains_(coil_dom), elec_attrs_(electrodes),
+      coef1_(nullptr), coef0_(nullptr), mesh_parent_(nullptr),
+      J_parent_(nullptr), V_parent_(nullptr), J_(nullptr), V_(nullptr),
+      high_src_(nullptr), low_src_(nullptr) {
+
+  coef1_ = new mfem::ConstantCoefficient(1.0);
+  coef0_ = new mfem::ConstantCoefficient(0.0);
```

Comment:
Some usage of 'naked new' here that could be avoided (via constructing the objects directly or via the use of smart pointers) - would make the class less fragile. Happy for this to be updated in a future PR though, given the expected updates to the `ClosedCoilSolver`

Reply:
You're right, I've changed the coefficients to be not be pointers anymore. I did the same to the high_src_ and low_src_ objects, which also didn't need to be pointers. I also changed some of the ScalarPotentialSource pointers to be deleted within the scope of the SPSCurrent function, such that they're not left hanging around with the risk of being initialised again. As for the remaining pointers in the class, I can try to turn them into shared pointers. 

I don't expect that the updates to `ClosedCoilSolver` will change much in `OpenCoilSolver`, so it's better to make `OpenCoilSolver` as robust as possible now. Feel free to point out as many fragilities in the class as you can find!